File: 2021/08/06/reverse_bits.py
"""
Problem: https://leetcode.com/problems/reverse-bits/
Author: Andrew Jarombek
Date: 8/6/2021
"""


def reverse_bits(n: int) -> int:
    binary = str(bin(n))[:1:-1]
    binary_suffix = "0" * (32 - len(binary))
    return int(binary + binary_suffix, 2)


if __name__ == '__main__':
    assert reverse_bits(43261596) == 964176192

File: 2023/03/26/cpp/reverseBits.cpp
// https://leetcode.com/problems/reverse-bits/
// Author: Andrew Jarombek
// Date: 3/26/2023

#include <cassert>
#include <iostream>

using namespace std;

uint32_t reverseBits(uint32_t n) {
    uint32_t result = 0;

    for (int i = 31; i >= 0; i--) {
        uint32_t val = n & 1;
        val <<= i;

        result += val;
        n >>= 1;
    }

    return result;
}

int main() {
    u_int32_t result = reverseBits(1);
    cout << result;
    assert(result == 2147483648);
}